lib/ansible/module_utils/identity/keycloak/keycloak.py

```diff
@@ -478,3 +480,134 @@ def delete_group(self, name=None, groupid=None, realm="master"):
 
         except Exception as e:
             self.module.fail_json(msg="Unable to delete group %s: %s" % (groupid, str(e)))
+
+    def get_users(self, realm="master"):
+        """ Fetch the name and ID of all users on the Keycloak server.
+
+        To fetch the full data of the user, make a subsequent call to
+        get_user_by_userid, passing in the ID of the user you wish to return.
+
+        :param realm: Return the users of this realm (default "master").
+        """
+        users_url = URL_USERS.format(url=self.baseurl, realm=realm)
+        try:
+            return json.loads(to_native(open_url(users_url, method="GET", headers=self.restheaders,
+                                                 validate_certs=self.validate_certs).read()))
+        except Exception as e:
+            self.module.fail_json(msg="Could not fetch list of users in realm %s: %s"
+                                      % (realm, str(e)))
+
+    def get_user_by_userid(self, gid, realm="master"):
+        """ Fetch a keycloak user from the provided realm using the user's unique ID.
+
+        If the user does not exist, None is returned.
+
+        gid is a UUID provided by the Keycloak API
+        :param gid: UUID of the user to be returned
+        :param realm: Realm in which the user resides; default 'master'.
+        """
+        users_url = URL_USER.format(url=self.baseurl, realm=realm, userid=gid)
+        try:
+            return json.loads(to_native(open_url(users_url, method="GET", headers=self.restheaders,
+                                                 validate_certs=self.validate_certs).read()))
+
+        except HTTPError as e:
+            if e.code == 404:
+                return None
+            else:
+                self.module.fail_json(msg="Could not fetch user %s in realm %s: %s"
+                                          % (gid, realm, str(e)))
+        except Exception as e:
+            self.module.fail_json(msg="Could not fetch user %s in realm %s: %s"
+                                      % (gid, realm, str(e)))
+
+    def get_user_by_username(self, username, realm="master"):
+        """ Fetch a keycloak user within a realm based on its username.
+
+        The Keycloak API does not allow filtering of the Groups resource by name.
+        As a result, this method first retrieves the entire list of users - username and ID -
+        then performs a second query to fetch the user.
+
+        If the user does not exist, None is returned.
+        :param username: Username of the user to fetch.
+        :param realm: Realm in which the user resides; default 'master'
+        """
+        users_url = URL_USERS.format(url=self.baseurl, realm=realm)
+        try:
+            all_users = self.get_users(realm=realm)
+
+            for user in all_users:
```

Comment:
It's possible to specify a username on the query string to the GET /users API call. The server will do filtering and return a reduced list of users. Could we use that instead of fetching down the full list to convert username to UUID?

It will save us bringing down large lists of users when we only need to select one.

This is something I need to implement in the keycloak_group module as well.

Reply:
I'll take a look
Thanks for your time @adamgoossens !

Reply:
ok it's done now. I now use the username filter.